ext/storage_sources/dir_store/dir_store.c

```diff
@@ -1096,55 +1096,57 @@ static int
+    ret = 0;
 
-    /* Skip over "./" and variations at the beginning of the name. */
-    while (*name == '.') {
-        if (name[1] != '/')
-            break;
-        name += 2;
-        while (*name == '/')
-            name++;
-    }
+    /* Check to see if the file exists in the cache directory before attempting to remove it. */
+    if ((ret = dir_store_cache_path(file_system, name, &cache_path)) != 0)
+        goto err;
+    ret = stat(cache_path, &sb);
 
-    /* Generate the paths to the shared object files in the bucket and cache directories. */
-    bucket_file_len = strlen(dir_store_fs->bucket_dir) + strlen(name) + 2;
-    if ((bucket_filename = calloc(bucket_file_len, sizeof(char))) == NULL)
-        return (dir_store_err(FS2DS(file_system), session, ENOMEM, "dir_store_remove"));
-    if (snprintf(bucket_filename, bucket_file_len, "%s/%s", dir_store_fs->bucket_dir, name) >=
-      (int)bucket_file_len)
-        return (dir_store_err(FS2DS(file_system), session, EINVAL, "overflow sprintf"));
-
-    cache_file_len = strlen(dir_store_fs->cache_dir) + strlen(name) + 2;
-    if ((cache_filename = calloc(cache_file_len, sizeof(char))) == NULL)
-        return (dir_store_err(FS2DS(file_system), session, ENOMEM, "dir_store_remove"));
-    if (snprintf(cache_filename, cache_file_len, "%s/%s", dir_store_fs->cache_dir, name) >=
-      (int)cache_file_len)
-        return (dir_store_err(FS2DS(file_system), session, EINVAL, "overflow sprintf"));
-
-    dir_store_exist(file_system, session, name, &existp);
-    if (existp) {
-        ret = wt_fs->fs_remove(wt_fs, session, bucket_filename, flags);
-        if (ret != 0 && ret != ENOENT)
+    if (ret != 0) {
+        if (errno != ENOENT) {
+            ret = dir_store_err(
+              dir_store_fs->dir_store, session, errno, "%s: dir_store_remove stat", cache_path);
+            goto err;
+        }
+    } else {
+        if ((ret = wt_fs->fs_remove(wt_fs, session, cache_path, flags)) != 0) {
+            ret = dir_store_err(
+              dir_store_fs->dir_store, session, errno, "%s: dir_store_remove", cache_path);
             goto err;
+        }
+    }
+
+    /* Check to see if the file exists in the bucket directory before attempting to remove it. */
+    if ((ret = dir_store_bucket_path(file_system, name, &bucket_path)) != 0)
+        goto err;
+    ret = stat(bucket_path, &sb);
 
-        ret = wt_fs->fs_remove(wt_fs, session, cache_filename, flags);
-        if (ret != 0 && ret != ENOENT)
+    if (ret != 0) {
+        if (errno != ENOENT) {
+            ret = dir_store_err(
+              dir_store_fs->dir_store, session, errno, "%s: dir_store_remove stat", bucket_path);
+            goto err;
+        }
+    } else {
+        if ((ret = wt_fs->fs_remove(wt_fs, session, bucket_path, flags)) != 0) {
+            ret = dir_store_err(
+              dir_store_fs->dir_store, session, errno, "%s: dir_store_remove", bucket_path);
             goto err;
+        }
```

Comment:
This is much better.  I'd still like to see this refactored - nearly identical code to check the bucket vs check the cache directory.  Make a private function to do the repeated work.
Sorry to keep beating on refactoring common bits, but I think always think it's an important element of code quality.